components/homme/src/share/prim_driver_base.F90

```diff
@@ -1468,113 +1404,114 @@ subroutine applyCAMforcing_tracers(elem,hvcoord,np1,np1_qdp,dt,adjustment)
+   else ! end of adjustment
+      ! apply forcing to Qdp
+      elem%derived%FQps(:,:)=0
+      do q=1,qsize
+         do k=1,nlev
             do j=1,np
                do i=1,np
-                  ! make Q consistent now that we have updated ps_v above
-                  ! recompute dp, since ps_v was changed above
-                  dp(i,j,k) = ( hvcoord%hyai(k+1) - hvcoord%hyai(k) )*hvcoord%ps0 + &
-                       ( hvcoord%hybi(k+1) - hvcoord%hybi(k) )*elem%state%ps_v(i,j,np1)
-                  elem%state%Q(i,j,k,ic)= &
-                       elem%state%Qdp(i,j,k,ic,np1_qdp)/dp(i,j,k)
-               end do
-            end do
-          end do
-        end do
-
-     else ! end of adjustment
-        ! apply forcing to Qdp
-        elem%derived%FQps(:,:)=0
-        do q=1,qsize
-           do k=1,nlev
-              do j=1,np
-                 do i=1,np
-                    v1 = dt*elem%derived%FQ(i,j,k,q)
-                    if (elem%state%Qdp(i,j,k,q,np1_qdp) + v1 < 0 .and. v1<0) then
-                       if (elem%state%Qdp(i,j,k,q,np1_qdp) < 0 ) then
-                          v1=0  ! Q already negative, dont make it more so
-                       else
-                          v1 = -elem%state%Qdp(i,j,k,q,np1_qdp)
-                       endif
-                    endif
-                    elem%state%Qdp(i,j,k,q,np1_qdp) = elem%state%Qdp(i,j,k,q,np1_qdp)+v1
-                    if (q==1) then
-                       elem%derived%FQps(i,j)=elem%derived%FQps(i,j)+v1/dt
-                    endif
-                 enddo
-              enddo
-           enddo
-        enddo
-
-        if (use_moisture) then
-           ! to conserve dry mass in the precese of Q1 forcing:
-           elem%state%ps_v(:,:,np1) = elem%state%ps_v(:,:,np1) + &
-             dt*elem%derived%FQps(:,:)
-        endif
-
-        ! Qdp(np1) and ps_v(np1) were updated by forcing - update Q(np1)
-        do k=1,nlev
-           dp(:,:,k) = ( hvcoord%hyai(k+1) - hvcoord%hyai(k) )*hvcoord%ps0 + &
-                ( hvcoord%hybi(k+1) - hvcoord%hybi(k))*elem%state%ps_v(:,:,np1)
-        enddo
-        do q=1,qsize
-           do k=1,nlev
-              elem%state%Q(:,:,k,q) = elem%state%Qdp(:,:,k,q,np1_qdp)/dp(:,:,k)
-           enddo
-        enddo
-
-     endif ! if adjustment
+                  v1 = dt*elem%derived%FQ(i,j,k,q)
+                  if (elem%state%Qdp(i,j,k,q,np1_qdp) + v1 < 0 .and. v1<0) then
+                     if (elem%state%Qdp(i,j,k,q,np1_qdp) < 0 ) then
+                        v1=0  ! Q already negative, dont make it more so
+                     else
+                        v1 = -elem%state%Qdp(i,j,k,q,np1_qdp)
+                     endif
+                  endif
+                  elem%state%Qdp(i,j,k,q,np1_qdp) = elem%state%Qdp(i,j,k,q,np1_qdp)+v1
+                  if (q==1) then
+                     elem%derived%FQps(i,j)=elem%derived%FQps(i,j)+v1/dt
+                     dp_adj(i,j,k)=dp_adj(i,j,k) + v1/dt
```

Comment:
Why do you normalize by dt here.  For FQps above it makes sense, since you will multiply by dt again when you update ps.  But, if "adjust_ps" is false below than dp_adj will be different but won't capture the timestep in the adjustment.

Reply:
good catch!   the adjust_ps=.false. capability will be new with this PR, and hasn't been used.  this would have been a very difficult bug to track down.

Reply:
pushed fix.